File: src/main/java/main/puzzle/assembly/dxz/zdd/ZDD.java
package main.puzzle.assembly.dxz.zdd;

import org.jetbrains.annotations.NotNull;

/**
 *
 * @author Bunnyspa
 */
public class ZDD {

    public static final ZDDNode TRUE_TERMINAL = new ZDDNode();

    @NotNull
    public static ZDDNode unique(int i, ZDDNode l, ZDDNode h, @NotNull ZDDNodeTable Z) {
        ZDDNode element = Z.get(i, l, h);
        if (element != null) {
            return element;
        }

        ZDDNode node = new ZDDNode(i, l, h);
        Z.add(node);
        return node;
    }
}

File: src/main/java/main/puzzle/assembly/dxz/zdd/ZDD.kt
package main.puzzle.assembly.dxz.zdd

/**
 *
 * @author Bunnyspa
 */
object ZDD {
    val TRUE_TERMINAL = ZDDNode()
    fun unique(i: Int, l: ZDDNode?, h: ZDDNode, Z: ZDDNodeTable): ZDDNode {
        val element = Z[i, l, h]
        if (element != null) {
            return element
        }
        val node = ZDDNode(i, l, h)
        Z.add(node)
        return node
    }
}